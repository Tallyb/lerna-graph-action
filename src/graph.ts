import PackageGraph from '@lerna/package-graph';
import Project from '@lerna/project';

export async function buildGraph(cwd: string): Promise<string> {
  const project = await new Project(cwd);
  if (!project.version) {
    throw new Error(`Cannot find project in ${cwd}`);
  }
  const packages = await project.getPackages();
  const graph = await new PackageGraph(packages);
  let dot = `digraph{`;

  for (const [, pkg] of graph) {
    for (const [, dep] of pkg.localDependencies) {
      //console.log('DEPP', packages.get(dep.name));
      dot += `"${pkg.name}\n${pkg.version}"->"${dep.name}\n${
        graph.get(dep.name).version
      }";`;
    }
  }
  dot += `}`;
  return dot;
}
