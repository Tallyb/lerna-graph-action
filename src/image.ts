import Axios from 'axios';
import qs from 'qs';
import Fs from 'fs';
export async function generateImage(
  graph: string,
  imagePath: string,
): Promise<void> {
  const writer = Fs.createWriteStream(imagePath);

  const data = qs.stringify({
    chl: graph,
    cht: 'gv',
  });
  const response = await Axios({
    url: 'https://image-charts.com/chart',
    method: 'POST',
    responseType: 'stream',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    data,
  });

  response.data.pipe(writer);

  return new Promise((resolve, reject) => {
    writer.on('finish', resolve);
    writer.on('error', reject);
  });
}
