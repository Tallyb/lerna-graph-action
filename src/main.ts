import {buildGraph} from './graph';
import {generateImage} from './image';
import Path from 'path';
import * as core from '@actions/core';

export async function run(): Promise<void> {
  try {
    const cwd: string =
      core.getInput('rootPath') || process.env.LERNA_ROOT || process.cwd();
    core.debug(`Creating a graph on ${cwd}`);
    const imagePath = Path.resolve(process.cwd(), core.getInput('imagePath'));
    const graph = await buildGraph(cwd);
    await generateImage(graph, imagePath);
    core.debug(`Creating image ${cwd}`);
  } catch (error) {
    core.setFailed(error.message);
  }
}
run();
