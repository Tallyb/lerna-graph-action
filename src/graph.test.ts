import {buildGraph} from '../src/graph';

describe('graph', () => {
  test('generate graph', async () => {
    const res = await buildGraph('__fixtures__/normal');
    expect(res).toMatchSnapshot();
  });
});
