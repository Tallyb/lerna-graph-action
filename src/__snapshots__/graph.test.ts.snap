// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph generate graph 1`] = `
"digraph{\\"package-2
1.0.0\\"->\\"package-1
1.0.0\\";\\"package-3
1.0.0\\"->\\"package-2
1.0.0\\";\\"package-5
1.0.0\\"->\\"package-3
1.0.0\\";\\"package-5
1.0.0\\"->\\"package-1
1.0.0\\";}"
`;
